demos/guide/quic-server-block.c

```diff
@@ -0,0 +1,299 @@
+    if (!SSL_set_fd(listener, fd))
+        goto err;
+
+    /* Begin listening. */
+    if (!SSL_listen(listener))
+        goto err;
+
+    /*
+     * Begin an infinite loop of listening for connections. We will only
+     * exit this loop if we encounter an error.
+     */
+    for (;;) {
+        /* Pristine error stack for each new connection */
+        ERR_clear_error();
+
+        /* Block while waiting for a client connection */
+        printf("Waiting for connection\n");
+        conn = SSL_accept_connection(listener, 0);
+        if (conn == NULL) {
+            fprintf(stderr, "error while accepting connection\n");
+            goto err;
+        }
+        printf("Accepted new connection\n");
+
+        /* Echo client input */
+        while (SSL_read_ex(conn, buf, sizeof(buf), &nread) > 0) {
+            if (SSL_write_ex(conn, buf, nread, &nwritten) > 0 &&
+                nwritten == nread) {
+                break;
+            }
+            fprintf(stderr, "Error echoing client input");
+            break;
+        }
+
+        /* Signal the end of the stream. */
+        if (SSL_stream_conclude(conn, 0) != 1) {
+            fprintf(stderr, "Unable to conclude stream\n");
+            SSL_free(conn);
+            goto err;
+        }
+
+        /*
+         * Shut down the connection. We may need to call this multiple times
+         * to ensure the connection is shutdown completely.
+         */
+        SSL_SHUTDOWN_EX_ARGS shutdown_args = {0};
+        while (SSL_shutdown_ex(conn, 0, &shutdown_args,
+                               sizeof(SSL_SHUTDOWN_EX_ARGS)) != 1) {
+            fprintf(stderr, "Re-attempting SSL shutdown\n");
+        }
+
+        SSL_free(conn);
+    }
+
+err:
+    SSL_free(listener);
+    return ok;
+}
+
+/* Minimal QUIC HTTP/1.0 server. */
+int main(int argc, char *argv[])
+{
+    int res = EXIT_FAILURE;
+    SSL_CTX *ctx = NULL;
+    int fd;
+    unsigned long port;
+
+#ifdef _WIN32
+    static const char *progname;
+    progname = argv[0];
+#endif
+
+    if (argc != 4) {
+        errx(res, "usage: %s <port> <server.crt> <server.key>", argv[0]);
+    }
+
+    /* Parse port number from command line arguments. */
+    port = strtoul(argv[1], NULL, 0);
+    if (port == 0 || port > UINT16_MAX) {
+        errx(res, "Failed to parse port number");
+    }
+
+    /* Create SSL_CTX that supports QUIC. */
+    if ((ctx = create_ctx(argv[2], argv[3])) == NULL) {
+        ERR_print_errors_fp(stderr);
+        errx(res, "Failed to create context");
+    }
+
+    /* Create and bind a UDP socket. */
+    if ((fd = create_socket((uint16_t)port)) < 0) {
+        SSL_CTX_free(ctx);
+        ERR_print_errors_fp(stderr);
+        errx(res, "Failed to create socket");
+    }
+
+    /* QUIC server connection acceptance loop. */
+    if (!run_quic_server(ctx, fd)) {
+        SSL_CTX_free(ctx);
+        ERR_print_errors_fp(stderr);
+        errx(res, "Error in QUIC server loop");
+    }
+
```

Comment:
It's a nit: we should also free `ctx` here because we do it in error branches.
I also think file descriptor `fd` should be closed explicitly just to make
people aware the descriptor is not closed when `ctx` is freed.

The `SSL_set_fd()` uses `BIO_set_fd(bio, fd, BIO_NOCLOSE);` so caller to
`SSL_set_fd()` is still responsible for file descriptor. So it should be also closed
before programm exits.

thanks.